Use text classification to work with the categories column

In [2]:
import pandas as pd
books = pd.read_csv("books_cleaned.csv")
books["categories"].value_counts().reset_index()

categories  count
0                      Fiction   2111
1             Juvenile Fiction    390
2    Biography & Autobiography    311
3                      History    207
4           Literary Criticism    124
..                         ...    ...
474               Conspiracies      1
475       Brothers and sisters      1
476             Rock musicians      1
477             Community life      1
478    Indic fiction (English)      1

[479 rows x 2 columns]

In [3]:
## Take a deeper look into some of the categories
books[books["categories"] == "Juvenile Fiction"]

isbn13      isbn10  \
30    9780006646006  000664600X   
79    9780020442608  0020442602   
85    9780030547744  0030547741   
86    9780060000141  0060000147   
87    9780060001537  0060001534   
...             ...         ...   
4845  9781590385814  1590385810   
4930  9781596792500  1596792507   
4942  9781599900056  159990005X   
5010  9781844580514  1844580512   
5157  9781932416374  1932416374   

                                                  title  \
30                                   Ocean Star Express   
79                       The voyage of the Dawn Treader   
85                             Where the Red Fern Grows   
86                                       Poppy's Return   
87                                    Diary of a Spider   
...                                                 ...   
4845                                         Fablehaven   
4930  Sherlock Holmes and the Case of the Hound of t...   
4942                                    The Drift House   
5010                               Attack of the Jaguar   
5157                The Very Persistent Gappers of Frip   

                                  authors        categories  \
30               Mark Haddon;Peter Sutton  Juvenile Fiction   
79                    Clive Staples Lewis  Juvenile Fiction   
85                           Wilson Rawls  Juvenile Fiction   
86                                    Avi  Juvenile Fiction   
87                          Doreen Cronin  Juvenile Fiction   
...                                   ...               ...   
4845                         Brandon Mull  Juvenile Fiction   
4930  Arthur Conan Doyle;Malvina G. Vogel  Juvenile Fiction   
4942                            Dale Peck  Juvenile Fiction   
5010                         M. A. Harvey  Juvenile Fiction   
5157           George Saunders;Lane Smith  Juvenile Fiction   

                                              thumbnail  \
30    http://books.google.com/books/content?id=I2QZA...   
79    http://books.google.com/books/content?id=fDD3C...   
85    http://books.google.com/books/content?id=IHpRw...   
86    http://books.google.com/books/content?id=XbcMJ...   
87    http://books.google.com/books/content?id=UWvZo...   
...                                                 ...   
4845  http://books.google.com/books/content?id=tbVIP...   
4930  http://books.google.com/books/content?id=EWgWP...   
4942  http://books.google.com/books/content?id=kbwPY...   
5010  http://books.google.com/books/content?id=3HUdt...   
5157  http://books.google.com/books/content?id=xYJOP...   

                                            description  published_year  \
30    Joe and his parents are enjoying a summer holi...          2002.0   
79    The "Dawn Treader" is the first ship Narnia ha...          1970.0   
85    A young boy living in the Ozarks achieves his ...          2000.0   
86    There's trouble at Gray House, the girlhood ho...          2006.0   
87    This is the diary ... of a spider. But don't b...          2005.0   
...                                                 ...             ...   
4845  When Kendra and Seth go to stay at their grand...          2006.0   
4930  Sherlock Holmes and Dr. Watson travel to the b...          2005.0   
4942  Sent to stay with their uncle in a ship-like h...          2006.0   
5010  This training manual for operatives of Xtreme ...          2004.0   
5157  A novel set in the seaside village of Frip fol...          2006.0   

      average_rating  num_pages  ratings_count  \
30              3.50       32.0            1.0   
79              4.09      216.0         2869.0   
85              4.37      288.0           95.0   
86              3.99      256.0         1086.0   
87              4.25       40.0         7903.0   
...              ...        ...            ...   
4845            4.09      351.0       111896.0   
4930            4.51      237.0           28.0   
4942            3.64      437.0          595.0   
5010            

In [4]:
## We may combine these specific categories as fiction vs. non fiction
books["categories"]

0                             Fiction
1       Detective and mystery stories
2                             Fiction
3                      Christian life
4                      Christian life
                    ...              
5192          Indic fiction (English)
5193                Adventure stories
5194                 Health & Fitness
5195                       Philosophy
5196                          History
Name: categories, Length: 5197, dtype: object

Mapping the top 10 categories as either Fiction, Nonfiction, Children's Fiction, Children's Nonfiction


In [5]:
category_mapping = {
    "Fiction" : "Fiction",
    "Juvenile Fiction" : "Children's Fiction",
    "Biography & Autobiography" : "Nonfiction",
    "History" : "Nonfiction",
    "Literary Criticism" : "Nonfiction",
    "Philosophy" : "Nonfiction",
    "Religion" : "Nonfiction",
    "Comics & Graphic Novels" : "Fiction",
    "Drama" : "Fiction",
    "Juvenile Nonfiction" : "Children's Nonfiction",
    "Science" : "Nonfiction",
    "Poetry" : "Nonfiction"
}

## Making a new simple column that maps top 10 categories to 4 specific categories
books['simple_category'] = books['categories'].map(category_mapping)
books

isbn13      isbn10  \
0     9780002005883  0002005883   
1     9780002261982  0002261987   
2     9780006178736  0006178731   
3     9780006280897  0006280897   
4     9780006280934  0006280935   
...             ...         ...   
5192  9788172235222  8172235224   
5193  9788173031014  8173031010   
5194  9788179921623  817992162X   
5195  9788185300535  8185300534   
5196  9789027712059  9027712050   

                                                  title  \
0                                                Gilead   
1                                          Spider's Web   
2                                        Rage of angels   
3                                        The Four Loves   
4                                   The Problem of Pain   
...                                                 ...   
5192                                  Mistaken Identity   
5193                                Journey to the East   
5194  The Monk Who Sold His Ferrari: A Fable About F...   
5195                                          I Am that   
5196                           The Berlin Phenomenology   

                                           authors  \
0                               Marilynne Robinson   
1                  Charles Osborne;Agatha Christie   
2                                   Sidney Sheldon   
3                              Clive Staples Lewis   
4                              Clive Staples Lewis   
...                                            ...   
5192                              Nayantara Sahgal   
5193                                 Hermann Hesse   
5194                                  Robin Sharma   
5195  Sri Nisargadatta Maharaj;Sudhakar S. Dikshit   
5196                 Georg Wilhelm Friedrich Hegel   

                         categories  \
0                           Fiction   
1     Detective and mystery stories   
2                           Fiction   
3                    Christian life   
4                    Christian life   
...                             ...   
5192        Indic fiction (English)   
5193              Adventure stories   
5194               Health & Fitness   
5195                     Philosophy   
5196                        History   

                                              thumbnail  \
0     http://books.google.com/books/content?id=KQZCP...   
1     http://books.google.com/books/content?id=gA5GP...   
2     http://books.google.com/books/content?id=FKo2T...   
3     http://books.google.com/books/content?id=XhQ5X...   
4     http://books.google.com/books/content?id=Kk-uV...   
...                                                 ...   
5192  http://books.google.com/books/content?id=q-tKP...   
5193  http://books.google.com/books/content?id=rq6JP...   
5194  http://books.google.com/books/content?id=c_7mf...   
5195  http://books.google.com/books/content?id=Fv_JP...   
5196  http://books.google.com/books/content?id=Vy7Sk...   

                                            description  published_year  \
0     A NOVEL THAT READERS and critics have been eag...          2004.0   
1     A new 'Christie for Christmas' -- a full-lengt...          2000.0   
2     A memorable, mesmerizing heroine Jennifer -- b...          1993.0   
3     Lewis' work on the nature of love divides love...          2002.0   
4     "In The Problem of Pain, C.S. Lewis, one of th...          2002.0   
...                                                 ...             ...   
5192  On A Train Journey Home To North India After L...          2003.0   
5193  This book tells the tale of a man who goes on ...          2002.0   
5194  Wisdom to Create a Life of Passion, Purpose, a...          2003.0   
5195  This collection of the timeless teachings of o...          1999.0   
5196  Since the three volume edition ofHegel's Philo...          1981.0   

      average_rating  num_pages  ratings_count  \
0               3.85      247.0          361.0   
1               3.83      241.0         5164.0   
2    

In [5]:
books.simple_category.isna().value_counts()

simple_category
False    3743
True     1454
Name: count, dtype: int64

We can see that 3743 of the total books have a simple category and 1454 of the books are without a simple category


In [6]:
## Use Zero-Shot Classification from HuggingFace to classify categories into simple categories
from transformers import pipeline

fiction_categories = ["Fiction", "Nonfiction"]

pipe = pipeline("zero-shot-classification",
                model="facebook/bart-large-mnli",
                device="cpu")

/Users/hanin/PycharmProjects/PythonProject/.venv/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
Device set to use cpu


In [7]:
## Show descriptions of simple category = fiction
books.loc[books["simple_category"] == "Fiction", "description"].reset_index(drop=True)[0]

## Make a sequence of text that we will pass through the zero-shot classifier
sequence = books.loc[books["simple_category"] == "Fiction", "description"].reset_index(drop=True)[0]

In [8]:
## Passing through the classifier
pipe(sequence, fiction_categories)

{'sequence': 'A NOVEL THAT READERS and critics have been eagerly anticipating for over a decade, Gilead is an astonishingly imagined story of remarkable lives. John Ames is a preacher, the son of a preacher and the grandson (both maternal and paternal) of preachers. It’s 1956 in Gilead, Iowa, towards the end of the Reverend Ames’s life, and he is absorbed in recording his family’s story, a legacy for the young son he will never see grow up. Haunted by his grandfather’s presence, John tells of the rift between his grandfather and his father: the elder, an angry visionary who fought for the abolitionist cause, and his son, an ardent pacifist. He is troubled, too, by his prodigal namesake, Jack (John Ames) Boughton, his best friend’s lost son who returns to Gilead searching for forgiveness and redemption. Told in John Ames’s joyous, rambling voice that finds beauty, humour and truth in the smallest of life’s details, Gilead is a song of celebration and acceptance of the best and the worst

This gives us a score of the probability of whether it is fiction or not. In this particular case, probability of Fiction is much higher


In [9]:
## We try to extract the label with the higher probability
import numpy as np
max_index = np.argmax(pipe(sequence, fiction_categories)["scores"]) ## gives us the score of the higher probability
max_label = pipe(sequence, fiction_categories)["labels"][max_index] ## gives us the label of the higher score
max_label

'Fiction'

In [12]:
## We now generalize and create a function that does the same
def generate_predictions(sequence, categories):
    predictions = pipe(sequence, categories)
    max_index = np.argmax(predictions["scores"])
    max_label = predictions["labels"][max_index]
    return max_label

actual_categories = []
predicted_categories = []


In [14]:
## Test the model

from tqdm import tqdm

## For the fiction category
for i in tqdm(range(0, 300)):
    sequence = books.loc[books["simple_category"] == "Fiction", "description"].reset_index(drop=True)[i]
    predicted_categories += [generate_predictions(sequence, fiction_categories)]
    actual_categories += ["Fiction"]

100%|██████████| 300/300 [03:58<00:00,  1.26it/s]


In [15]:
## For the Nonfiction category
for i in tqdm(range(0, 300)):
    sequence = books.loc[books["simple_category"] == "Nonfiction", "description"].reset_index(drop=True)[i]
    predicted_categories += [generate_predictions(sequence, fiction_categories)]
    actual_categories += ["Nonfiction"]

100%|██████████| 300/300 [04:19<00:00,  1.16it/s]


In [16]:
## Put both the actual and predicted categories into a dataframe to compare them
predictions_df = pd.DataFrame({"actual_cat": actual_categories, "predicted_cat": predicted_categories})
predictions_df

actual_cat predicted_cat
0       Fiction       Fiction
1       Fiction       Fiction
2       Fiction       Fiction
3       Fiction    Nonfiction
4       Fiction       Fiction
..          ...           ...
595  Nonfiction    Nonfiction
596  Nonfiction    Nonfiction
597  Nonfiction    Nonfiction
598  Nonfiction    Nonfiction
599  Nonfiction    Nonfiction

[600 rows x 2 columns]

Pretty close!


In [17]:
## To see how many of the predictions were correct
predictions_df["correct_predictions"] = (
    np.where(predictions_df["actual_cat"] == predictions_df["predicted_cat"], 1, 0)
)
correct_percent = predictions_df.correct_predictions.sum()/len(predictions_df)
correct_percent

np.float64(0.7783333333333333)

We see that the 77% of the predictions matched the actual predictions

In [18]:
## Now we may use this model to predict the categories of the missing categories
## First we extract the description and isbn of books with missing categories
isbns = []
predicted_cats = []

missing_cats = books.loc[books.simple_category.isna(), ["isbn13", "description"]].reset_index(drop=True)
missing_cats

isbn13                                        description
0     9780002261982  A new 'Christie for Christmas' -- a full-lengt...
1     9780006280897  Lewis' work on the nature of love divides love...
2     9780006280934  "In The Problem of Pain, C.S. Lewis, one of th...
3     9780006380832  Until Vasco da Gama discovered the sea-route t...
4     9780006470229  A new-cover reissue of the fourth book in the ...
...             ...                                                ...
1449  9788125026600  Not only does Nietzsche for Beginners delve in...
1450  9788171565641  Forster's lively, informed originality and wit...
1451  9788172235222  On A Train Journey Home To North India After L...
1452  9788173031014  This book tells the tale of a man who goes on ...
1453  9788179921623  Wisdom to Create a Life of Passion, Purpose, a...

[1454 rows x 2 columns]

In [19]:
## Then we try to predict the categories of these missing books using the above model that we constructed

for i in tqdm(range(0, len(missing_cats))):
    sequence = missing_cats.loc[i, "description"]
    predicted_label = generate_predictions(sequence, fiction_categories)
    predicted_cats.append(predicted_label)
    isbns.append(missing_cats.loc[i, "isbn13"])

## This populates predicated_cats with the predictions of categories and isbns with its corresponding isbn

100%|██████████| 1454/1454 [18:48<00:00,  1.29it/s]


In [26]:
## Testing the two lists

print(isbns[0:10],len(isbns))

[np.int64(9780002261982), np.int64(9780006280897), np.int64(9780006280934), np.int64(9780006380832), np.int64(9780006470229), np.int64(9780006472612), np.int64(9780006483014), np.int64(9780006483892), np.int64(9780006483908), np.int64(9780006486145)] 1454


In [27]:
## Testing the two lists

print(predicted_cats[0:10],len(predicted_cats))

['Fiction', 'Nonfiction', 'Nonfiction', 'Nonfiction', 'Fiction', 'Nonfiction', 'Fiction', 'Fiction', 'Fiction', 'Fiction'] 1454


Looks like both, isbns and predicted_cat works like we want it to

In [28]:
## We make a dataframe with the predicted categories and their isbns
missing_predicted_df = pd.DataFrame({"isbn13": isbns, "predicted_categories": predicted_cats})
missing_predicted_df

isbn13 predicted_categories
0     9780002261982              Fiction
1     9780006280897           Nonfiction
2     9780006280934           Nonfiction
3     9780006380832           Nonfiction
4     9780006470229              Fiction
...             ...                  ...
1449  9788125026600           Nonfiction
1450  9788171565641              Fiction
1451  9788172235222              Fiction
1452  9788173031014           Nonfiction
1453  9788179921623              Fiction

[1454 rows x 2 columns]

In [29]:
## We now merge this missing dataframe to the original dataframe, which is easier because we have the isbns of the books in the missing dataframe

books = pd.merge(books, missing_predicted_df, on="isbn13", how="left")
books

isbn13      isbn10  \
0     9780002005883  0002005883   
1     9780002261982  0002261987   
2     9780006178736  0006178731   
3     9780006280897  0006280897   
4     9780006280934  0006280935   
...             ...         ...   
5192  9788172235222  8172235224   
5193  9788173031014  8173031010   
5194  9788179921623  817992162X   
5195  9788185300535  8185300534   
5196  9789027712059  9027712050   

                                                  title  \
0                                                Gilead   
1                                          Spider's Web   
2                                        Rage of angels   
3                                        The Four Loves   
4                                   The Problem of Pain   
...                                                 ...   
5192                                  Mistaken Identity   
5193                                Journey to the East   
5194  The Monk Who Sold His Ferrari: A Fable About F...   
5195                                          I Am that   
5196                           The Berlin Phenomenology   

                                           authors  \
0                               Marilynne Robinson   
1                  Charles Osborne;Agatha Christie   
2                                   Sidney Sheldon   
3                              Clive Staples Lewis   
4                              Clive Staples Lewis   
...                                            ...   
5192                              Nayantara Sahgal   
5193                                 Hermann Hesse   
5194                                  Robin Sharma   
5195  Sri Nisargadatta Maharaj;Sudhakar S. Dikshit   
5196                 Georg Wilhelm Friedrich Hegel   

                         categories  \
0                           Fiction   
1     Detective and mystery stories   
2                           Fiction   
3                    Christian life   
4                    Christian life   
...                             ...   
5192        Indic fiction (English)   
5193              Adventure stories   
5194               Health & Fitness   
5195                     Philosophy   
5196                        History   

                                              thumbnail  \
0     http://books.google.com/books/content?id=KQZCP...   
1     http://books.google.com/books/content?id=gA5GP...   
2     http://books.google.com/books/content?id=FKo2T...   
3     http://books.google.com/books/content?id=XhQ5X...   
4     http://books.google.com/books/content?id=Kk-uV...   
...                                                 ...   
5192  http://books.google.com/books/content?id=q-tKP...   
5193  http://books.google.com/books/content?id=rq6JP...   
5194  http://books.google.com/books/content?id=c_7mf...   
5195  http://books.google.com/books/content?id=Fv_JP...   
5196  http://books.google.com/books/content?id=Vy7Sk...   

                                            description  published_year  \
0     A NOVEL THAT READERS and critics have been eag...          2004.0   
1     A new 'Christie for Christmas' -- a full-lengt...          2000.0   
2     A memorable, mesmerizing heroine Jennifer -- b...          1993.0   
3     Lewis' work on the nature of love divides love...          2002.0   
4     "In The Problem of Pain, C.S. Lewis, one of th...          2002.0   
...                                                 ...             ...   
5192  On A Train Journey Home To North India After L...          2003.0   
5193  This book tells the tale of a man who goes on ...          2002.0   
5194  Wisdom to Create a Life of Passion, Purpose, a...          2003.0   
5195  This collection of the timeless teachings of o...          1999.0   
5196  Since the three volume edition ofHegel's Philo...          1981.0   

      average_rating  num_pages  ratings_count  \
0               3.85      247.0          361.0   
1               3.83      241.0         5164.0   
2    

In [33]:
## Now we amend the simple categories to so that if they are empty, they take the values from the predicted category

books.simple_category = np.where(books.simple_category.isna(), books.predicted_categories, books.simple_category)
books

isbn13      isbn10  \
0     9780002005883  0002005883   
1     9780002261982  0002261987   
2     9780006178736  0006178731   
3     9780006280897  0006280897   
4     9780006280934  0006280935   
...             ...         ...   
5192  9788172235222  8172235224   
5193  9788173031014  8173031010   
5194  9788179921623  817992162X   
5195  9788185300535  8185300534   
5196  9789027712059  9027712050   

                                                  title  \
0                                                Gilead   
1                                          Spider's Web   
2                                        Rage of angels   
3                                        The Four Loves   
4                                   The Problem of Pain   
...                                                 ...   
5192                                  Mistaken Identity   
5193                                Journey to the East   
5194  The Monk Who Sold His Ferrari: A Fable About F...   
5195                                          I Am that   
5196                           The Berlin Phenomenology   

                                           authors  \
0                               Marilynne Robinson   
1                  Charles Osborne;Agatha Christie   
2                                   Sidney Sheldon   
3                              Clive Staples Lewis   
4                              Clive Staples Lewis   
...                                            ...   
5192                              Nayantara Sahgal   
5193                                 Hermann Hesse   
5194                                  Robin Sharma   
5195  Sri Nisargadatta Maharaj;Sudhakar S. Dikshit   
5196                 Georg Wilhelm Friedrich Hegel   

                         categories  \
0                           Fiction   
1     Detective and mystery stories   
2                           Fiction   
3                    Christian life   
4                    Christian life   
...                             ...   
5192        Indic fiction (English)   
5193              Adventure stories   
5194               Health & Fitness   
5195                     Philosophy   
5196                        History   

                                              thumbnail  \
0     http://books.google.com/books/content?id=KQZCP...   
1     http://books.google.com/books/content?id=gA5GP...   
2     http://books.google.com/books/content?id=FKo2T...   
3     http://books.google.com/books/content?id=XhQ5X...   
4     http://books.google.com/books/content?id=Kk-uV...   
...                                                 ...   
5192  http://books.google.com/books/content?id=q-tKP...   
5193  http://books.google.com/books/content?id=rq6JP...   
5194  http://books.google.com/books/content?id=c_7mf...   
5195  http://books.google.com/books/content?id=Fv_JP...   
5196  http://books.google.com/books/content?id=Vy7Sk...   

                                            description  published_year  \
0     A NOVEL THAT READERS and critics have been eag...          2004.0   
1     A new 'Christie for Christmas' -- a full-lengt...          2000.0   
2     A memorable, mesmerizing heroine Jennifer -- b...          1993.0   
3     Lewis' work on the nature of love divides love...          2002.0   
4     "In The Problem of Pain, C.S. Lewis, one of th...          2002.0   
...                                                 ...             ...   
5192  On A Train Journey Home To North India After L...          2003.0   
5193  This book tells the tale of a man who goes on ...          2002.0   
5194  Wisdom to Create a Life of Passion, Purpose, a...          2003.0   
5195  This collection of the timeless teachings of o...          1999.0   
5196  Since the three volume edition ofHegel's Philo...          1981.0   

      average_rating  num_pages  ratings_count  \
0               3.85      247.0          361.0   
1               3.83      241.0         5164.0   
2    

We see that the simple category is now fully populated, there are no empty values there.

In [36]:
## We may now drop the predicted categories column since we do not have any use for it anymore

books = books.drop(columns = ["predicted_categories"])
books

isbn13      isbn10  \
0     9780002005883  0002005883   
1     9780002261982  0002261987   
2     9780006178736  0006178731   
3     9780006280897  0006280897   
4     9780006280934  0006280935   
...             ...         ...   
5192  9788172235222  8172235224   
5193  9788173031014  8173031010   
5194  9788179921623  817992162X   
5195  9788185300535  8185300534   
5196  9789027712059  9027712050   

                                                  title  \
0                                                Gilead   
1                                          Spider's Web   
2                                        Rage of angels   
3                                        The Four Loves   
4                                   The Problem of Pain   
...                                                 ...   
5192                                  Mistaken Identity   
5193                                Journey to the East   
5194  The Monk Who Sold His Ferrari: A Fable About F...   
5195                                          I Am that   
5196                           The Berlin Phenomenology   

                                           authors  \
0                               Marilynne Robinson   
1                  Charles Osborne;Agatha Christie   
2                                   Sidney Sheldon   
3                              Clive Staples Lewis   
4                              Clive Staples Lewis   
...                                            ...   
5192                              Nayantara Sahgal   
5193                                 Hermann Hesse   
5194                                  Robin Sharma   
5195  Sri Nisargadatta Maharaj;Sudhakar S. Dikshit   
5196                 Georg Wilhelm Friedrich Hegel   

                         categories  \
0                           Fiction   
1     Detective and mystery stories   
2                           Fiction   
3                    Christian life   
4                    Christian life   
...                             ...   
5192        Indic fiction (English)   
5193              Adventure stories   
5194               Health & Fitness   
5195                     Philosophy   
5196                        History   

                                              thumbnail  \
0     http://books.google.com/books/content?id=KQZCP...   
1     http://books.google.com/books/content?id=gA5GP...   
2     http://books.google.com/books/content?id=FKo2T...   
3     http://books.google.com/books/content?id=XhQ5X...   
4     http://books.google.com/books/content?id=Kk-uV...   
...                                                 ...   
5192  http://books.google.com/books/content?id=q-tKP...   
5193  http://books.google.com/books/content?id=rq6JP...   
5194  http://books.google.com/books/content?id=c_7mf...   
5195  http://books.google.com/books/content?id=Fv_JP...   
5196  http://books.google.com/books/content?id=Vy7Sk...   

                                            description  published_year  \
0     A NOVEL THAT READERS and critics have been eag...          2004.0   
1     A new 'Christie for Christmas' -- a full-lengt...          2000.0   
2     A memorable, mesmerizing heroine Jennifer -- b...          1993.0   
3     Lewis' work on the nature of love divides love...          2002.0   
4     "In The Problem of Pain, C.S. Lewis, one of th...          2002.0   
...                                                 ...             ...   
5192  On A Train Journey Home To North India After L...          2003.0   
5193  This book tells the tale of a man who goes on ...          2002.0   
5194  Wisdom to Create a Life of Passion, Purpose, a...          2003.0   
5195  This collection of the timeless teachings of o...          1999.0   
5196  Since the three volume edition ofHegel's Philo...          1981.0   

      average_rating  num_pages  ratings_count  \
0               3.85      247.0          361.0   
1               3.83      241.0         5164.0   
2    

Now we have successfully merged the predicted categories of the missing categories to the simple categories and dropped the extra column

In [37]:
## We now save the books with predicted categories as a new file
books.to_csv("books_with_categories.csv", index=False)